shadowsocks/dialer.go

```diff
@@ -0,0 +1,165 @@
+	// `address` has the form `host:port`, where `host` can be a domain name or IP address.
+	DialUDP(address string) (onet.PacketConn, error)
+}
+
+type ssDialer struct {
+	proxyIP   net.IP
+	proxyPort int
+	cipher    shadowaead.Cipher
+}
+
+// NewDialer creates a Dialer that routes connections to a Shadowsocks proxy listening at
+// `host:port`, with authentication parameters `cipher` (AEAD) and `password`.
+func NewDialer(host, password, cipher string, port int) (Dialer, error) {
+	proxyIP, err := net.ResolveIPAddr("ip", host)
+	if err != nil {
+		return nil, errors.New("Failed to resolve proxy address")
+	}
+	aead, err := newAeadCipher(cipher, password)
+	if err != nil {
+		return nil, err
+	}
+	d := ssDialer{proxyIP: net.ParseIP(proxyIP.String()), proxyPort: port, cipher: aead}
+	return &d, nil
+}
+
+func (d *ssDialer) DialTCP(address string) (onet.DuplexConn, error) {
+	proxyAddr := &net.TCPAddr{IP: d.proxyIP, Port: d.proxyPort}
+	conn, err := net.DialTCP("tcp", nil, proxyAddr)
+	if err != nil {
+		return nil, err
+	}
+	ssw := NewShadowsocksWriter(conn, d.cipher)
+	socksTargetAddr := socks.ParseAddr(address)
+	if socksTargetAddr == nil {
+		return nil, errors.New("Invalid target address")
+	}
+	_, err = ssw.Write(socksTargetAddr)
+	if err != nil {
+		conn.Close()
+		return nil, errors.New("Failed to write target address")
+	}
+	ssr := NewShadowsocksReader(conn, d.cipher)
+	return onet.WrapConn(conn, ssr, ssw), nil
+}
+
+// Clients are encouraged to use io.ReadWriter methods of onet.PacketConn
+// to leverage the association with the proxy.
+func (d *ssDialer) DialUDP(address string) (onet.PacketConn, error) {
+	proxyAddr := &net.UDPAddr{IP: d.proxyIP, Port: d.proxyPort}
+	pc, err := net.ListenPacket("udp", "")
+	if err != nil {
+		return nil, err
+	}
+	targetHost, targetPortStr, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, errors.New("Invalid target address")
+	}
+	// Ignore the error, it would have failed when splitting the address
+	targetPort, _ := strconv.Atoi(targetPortStr)
+	targetAddr := &packetConnAddr{Host: targetHost, Port: targetPort}
+	conn := packetConn{
+		PacketConn: pc, proxyAddr: proxyAddr, targetAddr: targetAddr,
+			cipher: d.cipher, buf: make([]byte, udpBufSize)}
+	return &conn, nil
+}
+
+type packetConn struct {
+	net.PacketConn
+	io.ReadWriter
+	proxyAddr  *net.UDPAddr
+	targetAddr net.Addr
+	cipher     shadowaead.Cipher
+	m          sync.Mutex
+	buf        []byte // Write lock
+}
+
+func (c *packetConn) Write(b []byte) (int, error) {
+	return c.WriteTo(b, c.proxyAddr)
+}
+
+// WriteTo encrypts b and write to addr using the embedded PacketConn.
+func (c *packetConn) WriteTo(b []byte, addr net.Addr) (int, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	socksTargetAddr := socks.ParseAddr(c.targetAddr.String())
+	if socksTargetAddr == nil {
+		return 0, errors.New("Invalid target address")
+	}
+	buf, err := shadowaead.Pack(c.buf, append(socksTargetAddr, b...), c.cipher)
```

Comment:
append may do an allocation. Instead, maybe you can use copy into c.buf of the target address and the bytes to avoid the extra allocation.

It will be helpful to add a benchmark to count that.

Reply:
I experimented with copying into the `c.buf` and found two issues: `b` may not fit in`c.buf`, and passing `c.buf` to `Pack` as destination and cipher buffers may cause an buffer overlap error.

Reply:
In that case, maybe you should get _two_ buffers from the pool

Reply:
We now reuse the allocated slice as cipher and plaintext buffers. We resort to append (instead of copy) in case the user buffer exceeds the max UDP packet length.